Documentation/High Availability and Scale.md

```diff
@@ -0,0 +1,369 @@
+1. Both SDN Appliances are still available in case of single TOR
+    failure
+1. Some ENIs are "Active" on SDN Appliance 1, and "Passive" on SDN
+    Appliance 2. Other ENIs are "Active" on SDN Appliance 2 and
+    "Passive" on SDN Appliance 1
+1. Both SDN Appliances are in use for traffic
+1. Each SDN Appliance is provisioned only up to a set threshold of
+    capacity (to allow for failovers)
+
+    - We use 80% in this document and describe in more detail in
+        subsequent sections
+    - If interested in full throughput redundancy -- a party may
+        choose to run at 50%
+1. Single TOR failure does not impact CPS
+1. Single TOR failure reduces available bandwidth
+
+### Traffic flow in Highly Available and Scalable Sirius Appliance Set
+
+![traffic-flow-has-scale-appliance](images/has/traffic-flow-has-scale-appliance.png)
+
+**Considerations**
+
+1. ENIs from single VM provisioned on multiple cards on same SDN Appliance
+1. Card from SDN Appliance 1 is "paired" with card from SDN Appliance 2 (each card participates only in single pairing relationship)
+1. Pairwise flow replication
+1. Single card handles multiple ENIs
+1. Some ENIs on same card are in "Active" mode, some other ENIs on same card are in "Passive" mode
+1. Each card has two VIPs: one VIP announced with short path thru BGP (used by ENIs in "Active" mode), 
+second VIP announced with longer path thru BGP (used by ENIs in "Passive" mode)
+1. Paired cards announce same set of VIPs
+1. ENI-based (not card based) flow replication (direction of flow replication: "Active ENI" -\> "Passive ENI")
+1. Single ENI is programmed on multiple cards (each card with different VIP)
+1. TOR (or source node where VM is located) performs traffic load balancing/splitting/sharding for 
+selected ENI across VIPs of the cards on which this ENI is provisioned
+
+## High Availability Architecture
+
+### Active-Passive model
+
+Active-Passive model will ensure high availability in case either (1)
+one of the TOR/SDN Appliance dies, or (2) single card dies.
+
+**Setup details**
+
+1. Each card from "SDN Appliance 1" will have "paired" card from "SDN Appliance 2"
+1. "Paired" cards will be serving same ENI with exact same policies setup on each
+1. "Paired" cards will be continuously replicating active flows from Active card to Passive card
+1.  Both cards will be announcing same VIP via BGP
+1. "SDN Appliance 1" will be in Active mode (announcing preferred, shorter path to itself thru BGP)
+1. "SDN Appliance 2" will be in Passive mode (announcing less preferred, longer path to itself thru BGP)
+
+**Normal traffic pattern**
+
+Normal traffic pattern for ENIs handled by cards will be going always thru "SDN Appliance 1" (Active one).
+
+**Failure mode**
+
+In case of failure the BGP routes from "SDN Appliance 1" (previously active) will be withdrawn and TOR will prefer 
+"SDN Appliance 2" and redirect traffic there, ensuring continuous traffic and uninterrupted customer experience.
```

Comment:
Is the service truly uninterrupted? Isn't there a finite time delay to detect the failure, then withdraw routes? If there is a "short" interruption, we should state some objective, e.g. 90% probability of restoration in under XX time span.

Reply:
Question for @mzms.